exchange/utils.go

```diff
@@ -84,9 +84,26 @@ func cleanOpenRTBRequests(orig *openrtb.BidRequest, adapters []openrtb_ext.Bidde
 			newReq := new(openrtb.BidRequest)
 			// Make a shallow copy of the original request
 			*newReq = *orig
+			prepareUser(newReq, adapters[i], usersyncs)
 			newReq.Imp = newImps
 			cleanReqs[adapters[i]] = newReq
 		}
 	}
 	return cleanReqs, errList
 }
+
+// prepareUser changes req.User so that it's ready for the given bidder.
+// This *will* mutate the request, but will *not* mutate any objects nested inside it.
+func prepareUser(req *openrtb.BidRequest, bidder openrtb_ext.BidderName, usersyncs IdFetcher) {
+	if id, ok := usersyncs.GetId(bidder); ok {
+		if req.User == nil {
+			req.User = &openrtb.User{
+				BuyerUID: id,
+			}
+		} else if req.User.BuyerUID == "" {
+			clone := *req.User
+			clone.BuyerUID = id
+			req.User = &clone
```

Comment:
Curious, what happens if `req.User.BuyerUID=id` without `clone`? 

I'm guessing the comment about mutation has something to do with it, but didn't quite get why it's important to not modify the original User object.

Reply:
The BidRequest stores a _pointer_ to the User object. This code splits that original request into many... but without the clone, both BidRequests point to the _same_ User object.

So basically all the bidders would see the `BuyerUID` of the last one in our loop, rather than the one which belongs to them.